### ISSUE TITLE
extremly long startup time

### ISSUE THREAD

**Issue by RubenKelevra** [ipfs/go-ipfs#7845]
#### Version information:
go-ipfs version: 0.9.0-dev-2ed925442
Repo version: 11
System version: amd64/linux
Golang version: go1.15.6

#### Description:
I'm currently having a hard time starting ipfs on one of my servers. It will just print the following into the console on startup but won't come online:

```
# journalctl -f -b0 -u ipfs
-- Journal begins at Fri 2020-12-11 13:45:01 CET. --
Dez 29 06:27:36 odin.pacman.store systemd[1]: Starting InterPlanetary File System (IPFS) daemon...
Dez 29 06:27:36 odin.pacman.store ipfs[576]: Initializing daemon...
Dez 29 06:27:36 odin.pacman.store ipfs[576]: go-ipfs version: 0.9.0-dev-2ed925442
Dez 29 06:27:36 odin.pacman.store ipfs[576]: Repo version: 11
Dez 29 06:27:36 odin.pacman.store ipfs[576]: System version: amd64/linux
Dez 29 06:27:36 odin.pacman.store ipfs[576]: Golang version: go1.15.6
```

Since port 5001 is still closed while this happens, I've killed it with -SIGABRT, and here's the stack trace.

[ipfs_no_startup.log](https://github.com/ipfs/go-ipfs/files/5749707/ipfs_no_startup.log)

**Comment by aschmahmann** [comment]
@RubenKelevra is it possible your MFS root is corrupted? https://github.com/ipfs/go-ipfs/issues/7183

**Comment by RubenKelevra** [comment]
> @RubenKelevra is it possible your MFS root is corrupted? https://github.com/ipfs/go-ipfs/issues/7183

I mean, this bug only lasted until the next restart of the daemon.

It worked for some days fine and then got stuck again.

The shutdowns are not "gracefully" after it get stuck with the #7183.

IPFS runs with flatfs as datastore which (and all other ipfs data) is stored on a ZFS file system which showed no issues and the server itself never crashed since I've installed it.

To be sure that the file system is in good shape I did a scrub, which turned up no issues.

The server ran for 163 days with ipfs 0.6 without any issues, #7183 and this issue just appeared after the update to 0.8.

If anything is corrupted in the IPFS database than it's because of a bug in ipfs or a dependency introduced since 0.6 I guess.

**Comment by RubenKelevra** [comment]
I couldn't start-up the daemon in the 15-minute timeout window I set for the systemd service. So I increased it to infinity. I tried to start the daemon in the morning of the 16th of January, but it's still not started up (and reported back to systemd that it's running) more than 3 days later. :neutral_face: 

I killed it again with sigabrt to create a stack trace:

[ipfs_wont_finish_startup.log](https://github.com/ipfs/go-ipfs/files/5833164/ipfs_wont_finish_startup.log)

**Comment by aschmahmann** [comment]
@RubenKelevra as I mentioned above I'm pretty sure your MFS root is corrupted and that's why it's hanging. You can see from your log that your node is stuck on this line in the startup process https://github.com/ipfs/go-ipfs/blob/7588a6a52a789fa951e1c4916cee5c7a304912c2/core/node/core.go#L140.

Basically your node tries to load your MFS root on startup, and while it should be fetching the root in offline mode it's actually willing to search the network for that node and so is endlessly searching the network looking for an MFS root that nobody has.  This behavior means that if you're running multiple IPFS nodes and keeping their MFS's in sync then it's possible your corrupted node can actually recover the data from the non-corrupted node.

So where to move from here:
1) You should decide what you'd like to do about the state of your MFS. You can use the tools linked in that issue to manually override your MFS root (i.e. the more specific https://github.com/hsanjuan/mfs-replace-root or the more generic https://github.com/aschmahmann/ipfs-ds). You can either set it to just be the empty directory or if you know of a valid directory you'd like to set it to then you can do that. There's also the tool linked in #7183 for finding the largest DAG in your repo if that's a helpful recovery mechanism for you.
2) It seems like maybe you've been able to duplicate/reproduce this behavior. Any more details on what's been happening on the node that causes this? e.g. disk runs out of space, daemon killed while modifying mfs, simultaneous mfs operations, large mfs operations, etc. Anything that might help with diagnostic info here could be useful.
3) This is a duplicate of the above issue so we should close it and add relevant diagnostic information in that thread

Seem reasonable? If following the MFS root corruption instructions doesn't resolve your startup issue then we can reopen this issue as something different from #7183.

**Comment by RubenKelevra** [comment]
> Basically your node tries to load your MFS root on startup, and while it should be fetching the root in offline mode it's actually willing to search the network for that node and so is endlessly searching the network looking for an MFS root that nobody has. This behavior means that if you're running multiple IPFS nodes and keeping their MFS's in sync then it's possible your corrupted node can actually recover the data from the non-corrupted node.

Well, there are only local operations running on this node. So data gets added locally and then moved to the right locations in the mfs until the mfs is exactly how it should look like and then it's starting over again.

So there's no chance to recover with data from other nodes since the state wasn't yet replicated at this stage - it's still in the 'creating a new state' process.

> 1. You should decide what you'd like to do about the state of your MFS. You can use the tools linked in that issue to manually override your MFS root (i.e. the more specific https://github.com/hsanjuan/mfs-replace-root or the more generic https://github.com/aschmahmann/ipfs-ds). You can either set it to just be the empty directory or if you know of a valid directory you'd like to set it to then you can do that. There's also the tool linked in #7183 for finding the largest DAG in your repo if that's a helpful recovery mechanism for you.

Well, there's nothing noteworthy of data in this node. So I can just reimport the data if necessary. But I just want to make sure it doesn't happen again before I start :)

I already tried one of the tools you mentioned, and it's currently not working, see https://github.com/hsanjuan/mfs-replace-root/issues/1.

> 2. It seems like maybe you've been able to duplicate/reproduce this behavior. Any more details on what's been happening on the node that causes this? e.g. disk runs out of space, daemon killed while modifying mfs, simultaneous mfs operations, large mfs operations, etc. Anything that might help with diagnostic info here could be useful.

Yeah sure.

My script runs the following commands before it gets stuck:

```bash
ipfs files stat --hash /x86-64.archlinux.pkg.pacman.store/lastsync # which returned 0 - so there's a file
ipfs files rm /x86-64.archlinux.pkg.pacman.store/lastsync #which returns 0
ipfs add --chunker size-262144 --hash blake2b-256 --cid-version 1 --raw-leaves --trickle --quieter /var/lib/ipfs/x86-64.archlinux.pkg.pacman.store/lastsync # which returns 0 and a cid
ipfs api files cp /ipfs/bafykbzaceafqnjhgsdwwz3bm66y7ti7z6te4773vgh4pn4axiqoluxjgydmhc /x86-64.archlinux.pkg.pacman.store/lastsync # uses the cid of the previous command, but never returns
```

There are no concurrent operations running except maybe some triggered by IPFS-Cluster. But since the cluster is a '-1' redundancy it shouldn't really command any pins etc at that time.



> 3. This is a duplicate of the above issue so we should close it and add relevant diagnostic information in that thread
> 
> Seem reasonable? If following the MFS root corruption instructions doesn't resolve your startup issue then we can reopen this issue as something different from #7183.

Yeah sure. As soon as issue with mfs-replace-root meantioned above is fixed I can validate if this would fix the startup :)

**Comment by aschmahmann** [comment]
> But I just want to make sure it doesn't happen again before I start :)

Definitely hear you there, and appreciate the detailed steps that could be useful in debugging here. Running this script messes things up somewhat predictably, i.e. this has happened multiple times in the last few weeks?

> Yeah sure. As soon as issue with mfs-replace-root mentioned above is fixed I can validate if this would fix the startup :)

https://github.com/aschmahmann/ipfs-ds has its deps up to date if you don't want to wait on mfs-replace-root to be updated. The instructions for resetting the MFS root are in the README. Note the `z` in `ipfs-ds put --value-encoded /local/filesroot zQmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn`, it's there so that the program knows that Qm... should be decoded as base58btc before being stored as binary in the datastore.

**Comment by RubenKelevra** [comment]
Hey, @aschmahmann I feel like debugging this is a bit over my resources right now. I packaged the whole IPFS folder if you want to try it out locally and look if the MFS is the issue.

https://github.com/ipfs/go-ipfs/issues/7844#issuecomment-769082004
